### PR TITLE
Read property from 'CRUD' in Problem Model

### PR DESCRIPTION
**Pull request by ferdinandjason**

### CODE REVIEW

File: app/Api/V1/Controllers/ProblemController.php [hunk 1]
```diff
@@ -32,4 +35,29 @@ public function store(
         $service->createOne($params);
         return $this->response->created();
     }
+
+    public function index(
+        ProblemTransformer $problemTransformer,
+        ReadProblemService $service,
+        Request $request
+    ):Response
```
Comment (walbertus) [inline app/Api/V1/Controllers/ProblemController.php]: Missing `space` before Response
You should try to use `ctrl + shift + l` (or `ctrl + windows + shift + l` in linux)  to autoformat your code in phpstorm

File: app/Api/V1/Controllers/ProblemController.php [hunk 2]
```diff
@@ -32,4 +35,29 @@ public function store(
...
+    {
+        $fields = $request->only(ReadProblemParam::QUERY_INDEX_PARAMS);
+
+        $params = new ReadProblemParam();
+        $params->fromArray($fields);
+
+        $problems = $service->readMany($params);
+        return $this->response->item($problems,$problemTransformer);
```
Comment (walbertus) [inline app/Api/V1/Controllers/ProblemController.php]: This should return Collection Response (many items)

File: app/Api/V1/Domain/Problem/Param/CreateProblemParam.php [hunk 3]
```diff
@@ -17,7 +17,7 @@ class CreateProblemParam
     const QUERY_PARAM_VALIDATION = [
         Problem::ATTRIBUTE_SLUG => [
             'required',
-            'regex:/^[a-z0-9]+(?:-[a-z0-9]+)*$/',
+            'regex:/^[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*$/',
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Param/CreateProblemParam.php]: I think we need to use UPPERCASE only or lowercase only
I prefer UPPERCASE only, you can remove the lowercase character

File: app/Api/V1/Domain/Problem/Param/ReadProblemParam.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace App\Api\V1\Domain\Problem\Param;
+
+
+class ReadProblemParam
+{
+    const ATTRIBUTE_PAGE = 'page';
+    const ATTRIBUTE_LIMIT = 'limit';
+
+    const QUERY_INDEX_PARAMS = [
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Param/ReadProblemParam.php]: This should be `QUERY_PARAMS`

File: app/Api/V1/Domain/Problem/Repository/ProblemRepository.php [hunk 5]
```diff
@@ -3,11 +3,22 @@
...
+use Illuminate\Database\Eloquent\Collection;
 
 class ProblemRepository
 {
     public function createOne(array $data): Problem
     {
         return Problem::create($data);
     }
+
+    public function readMany($offset, $limit): Collection
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Repository/ProblemRepository.php]: You need to use `Type Hint` for input parameter

File: app/Api/V1/Domain/Problem/Repository/ProblemRepository.php [hunk 6]
```diff
@@ -3,11 +3,22 @@
...
+    {
+        return Problem::skip($offset)->take($limit)->get();
+    }
+
+    public function readSingle($id): Problem
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Repository/ProblemRepository.php]: You need to use `Type Hint` for input parameter

File: app/Api/V1/Domain/Problem/Services/ReadProblemService.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    public function __construct(ProblemRepository $repository)
+    {
+        $this->repository = $repository;
+    }
+
+    public function readMany(ReadProblemParam $param) : Collection
+    {
+        $param = $param->toArray();
+
+        $offset = $param[ReadProblemParam::ATTRIBUTE_PAGE]*$param[ReadProblemParam::ATTRIBUTE_LIMIT];
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Services/ReadProblemService.php]: To get parameter from Param object you should declare getter function (`getPage` and `getLimit` or `getOffset`)

File: app/Api/V1/Domain/Problem/Services/ReadProblemService.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        $limit = $param[ReadProblemParam::ATTRIBUTE_LIMIT];
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Services/ReadProblemService.php]: use `getter`

File: app/Api/V1/Domain/Problem/Services/ReadProblemService.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+        return $this->repository->readMany($offset,$limit);
+    }
+
+    public function readSingle($id) : Problem
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Services/ReadProblemService.php]: use `Type Hint` for input param

File: config/cors.php [hunk 10]
```diff
@@ -20,4 +20,4 @@
     'exposedHeaders' => [],
     'maxAge' => 0,
 
-];
+];
```
Comment (walbertus) [inline config/cors.php]: unnecessary change

**Review by walbertus — CHANGES_REQUESTED** [review]
Please update

File: app/Api/V1/Domain/Problem/Param/CreateProblemParam.php (hunk 3 above)
Comment (ferdinandjason) [reply app/Api/V1/Domain/Problem/Param/CreateProblemParam.php]: Roger that 🙆

File: app/Api/V1/Controllers/ProblemController.php (hunk 1 above)
Comment (ferdinandjason) [reply app/Api/V1/Controllers/ProblemController.php]: Autoformat in phpstroms is `ctrl + alt + shift + l` :+1:

File: app/Api/V1/Domain/Problem/Param/ReadProblemParam.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace App\Api\V1\Domain\Problem\Param;
+
+
+class ReadProblemParam
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Param/ReadProblemParam.php:6]: You should check validation for every input

in this case i think `limit` should be `numeric|min:0|nullable`

File: app/Api/V1/Domain/Problem/Param/ReadProblemParam.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        $this->data[ReadProblemParam::ATTRIBUTE_LIMIT] = null;
+    }
+
+    public function fromArray(array $array): void
+    {
+        $this->data[ReadProblemParam::ATTRIBUTE_LIMIT] = $array[ReadProblemParam::ATTRIBUTE_LIMIT];
+    }
+
+    public function getLimit(): int
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Param/ReadProblemParam.php]: Limit could be null so this function return type should not be `int`, you could use `?int`

**Review by walbertus — CHANGES_REQUESTED** [review]

File: app/Api/V1/Controllers/ProblemController.php [hunk 13]
```diff
@@ -32,4 +36,33 @@ public function store(
...
         return $this->response->created();
     }
+
+    public function index(
+        ProblemTransformer $problemTransformer,
+        ReadProblemService $service,
+        Request $request
+    ): Response
+    {
+        $limit = $request->get(self::QUERY_LIMIT, 10);
```
Comment (walbertus) [inline app/Api/V1/Controllers/ProblemController.php]: constant should use const static variable (`10`)
in `ProblemController` there should be a constant called `DEFAULT_LIMIT`
then `$request->get(self::QUERY_LIMIT, self::DEFAULT_LIMIT);`

File: app/Api/V1/Domain/Problem/Repository/ProblemRepository.php [hunk 14]
```diff
@@ -3,11 +3,26 @@
...
 class ProblemRepository
 {
     public function createOne(array $data): Problem
     {
         return Problem::create($data);
     }
+
+    public function readMany(?int $ownerId, $limit): Paginator
+    {
+        if ($ownerId) {
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Repository/ProblemRepository.php]: This doesn't look good
I think you should create BuildQuery function with queryParam input and return array of where condition

File: app/Api/V1/Domain/Problem/Helper/ProblemQueryGenerator.php [hunk 15]
```diff
@@ -0,0 +1,31 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Helper/ProblemQueryGenerator.php]: Ini masukin ke modul repository aja
sama comment diatas dihilangkan saja

File: app/Api/V1/Domain/Problem/Services/ReadProblemService.php [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    protected $repository;
+
+    public function __construct(ProblemRepository $repository)
+    {
+        $this->repository = $repository;
+    }
+
+    public function readMany(ReadProblemParam $param, int $limit): Paginator
+    {
+        $query = ProblemQueryGenerator::buildQuery($param);
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Services/ReadProblemService.php]: Bagian ini baiknya ada di repository karena menyangkut Database logic
Di service tugasnya application/business logic saja
Jadi pass query param nya ke repository

File: app/Api/V1/Domain/Problem/Helper/ProblemQueryGenerator.php [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+namespace App\Api\V1\Domain\Problem\Helper;
+
+
+use App\Api\V1\Domain\Problem\Param\ReadProblemParam;
+
+class ProblemQueryGenerator
+{
+    static public function buildQuery(ReadProblemParam $queryParams): array
```
Comment (walbertus) [inline app/Api/V1/Domain/Problem/Helper/ProblemQueryGenerator.php]: Sepertinya akan lebih baik untuk tidak menggunakan static function supaya lebih mudah dalam automation testing karena lebih mudah di mock

Comment (walbertus) [reply app/Api/V1/Domain/Problem/Helper/ProblemQueryGenerator.php]: Untuk penggunaan di class lain maka bisa di inisialisasi pada constructor nya

**Review by walbertus — APPROVED** [review]
